Modules/_winapi.c

```diff
@@ -744,6 +744,16 @@ getenvironment(PyObject* environment)
                 "environment can only contain strings");
             goto error;
         }
+        if (PyUnicode_FindChar(key, '\0', 0, PyUnicode_GET_LENGTH(key), 1) != -1 ||
+            PyUnicode_FindChar(value, '\0', 0, PyUnicode_GET_LENGTH(value), 1) != -1)
+        {
+            PyErr_SetString(PyExc_ValueError, "embedded null character");
+            goto error;
+        }
+        if (PyUnicode_FindChar(key, '=', 0, PyUnicode_GET_LENGTH(key), 1) != -1) {
```

Comment:
As far as I know you won't find anything helpful on MSDN. Their article on changing environment variables says that equals cannot be used in the name, without explaining its use as the first character. Also, their article on changing the current directory incorrectly states that `SetCurrentDirectory` remembers per-drive directories, which it definitely never does. 

The Windows API consumes these variables when resolving paths, but it never sets them. That's left to applications, and usually via the C runtime `chdir` function. Since Python has its own implementation of `chdir` on Windows, you can see how it sets these variables in [`win32_wchdir`](https://github.com/python/cpython/blob/a3d91b43c2851312fb942f31afa12f5961706db6/Modules/posixmodule.c#L1498).